tree/tree.go

```diff
@@ -0,0 +1,723 @@
+	t.value = root
+	t.isRoot = true
+	t.tree = ltree.Root(root)
+	return t
+}
+
+// New creates a new model with default settings.
+func New(t *Node, width, height int) Model {
+	m := Model{
+		KeyMap:          DefaultKeyMap,
+		OpenCharacter:   "▼",
+		ClosedCharacter: "▶",
+		Help:            help.New(),
+
+		showHelp: true,
+		root:     t,
+	}
+	m.SetStyles(DefaultStyles())
+	m.setSize(width, height)
+	m.setAttributes()
+	m.updateStyles()
+	return m
+}
+
+// Update is the Bubble Tea update loop.
+func (m Model) Update(msg tea.Msg) (Model, tea.Cmd) {
+	var cmds []tea.Cmd
+
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
+		switch {
+		case key.Matches(msg, m.KeyMap.Down):
+			m.yOffset = min(m.root.size-1, m.yOffset+1)
+		case key.Matches(msg, m.KeyMap.Up):
+			m.yOffset = max(0, m.yOffset-1)
+		case key.Matches(msg, m.KeyMap.Toggle):
+			node := findNode(m.root, m.yOffset)
+			if node == nil {
+				break
+			}
+			node.open = !node.open
+			if node.open {
+				node.tree.Offset(0, 0)
+			} else {
+				node.tree.Offset(0, node.tree.Children().Length())
+			}
+			m.setAttributes()
+			m.yOffset = node.yOffset
+		case key.Matches(msg, m.KeyMap.Quit):
+			return m, tea.Quit
+		case key.Matches(msg, m.KeyMap.ShowFullHelp):
+			fallthrough
+		case key.Matches(msg, m.KeyMap.CloseFullHelp):
+			m.Help.ShowAll = !m.Help.ShowAll
+		}
+	}
+
+	// not sure why, but I think m.yOffset is captured in the closure, so we need to update the styles
+	m.updateStyles()
+	return m, tea.Batch(cmds...)
+}
+
+// View renders the component.
+func (m Model) View() string {
+	var treeView, leftDebugView, cursor string
+	// TODO: remove
+	if os.Getenv("DEBUG") == "true" {
+		// topDebugView += fmt.Sprintf("y=%2d\n", m.yOffset)
+		leftDebugView = printDebugInfo(m.root) + " "
+		for i := 0; i < m.root.size; i++ {
+			if i == m.yOffset {
+				cursor = cursor + "👉 "
+			} else {
+				cursor = cursor + "  "
+			}
+			cursor = cursor + "\n"
+		}
+	}
+
+	treeView = m.styles.TreeStyle.Render(m.root.String())
+
+	treeView = lipgloss.JoinHorizontal(
+		lipgloss.Top,
+		leftDebugView,
+		cursor,
+		treeView,
+	)
+
+	var help string
+	if m.showHelp {
+		help = m.helpView()
+	}
+
+	return lipgloss.JoinVertical(lipgloss.Left, treeView, help)
+}
+
+// SetStyles sets the styles for this component.
+func (m *Model) SetStyles(styles Styles) {
+	if styles.NodeStyleFunc != nil {
+		styles.nodeFunc = styles.NodeStyleFunc
+	} else {
+		styles.nodeFunc = func(_ Nodes, _ int) lipgloss.Style {
+			return styles.NodeStyle
+		}
+	}
+	if styles.SelectedNodeStyleFunc != nil {
+		styles.selectedNodeFunc = styles.SelectedNodeStyleFunc
+	} else {
+		styles.selectedNodeFunc = func(_ Nodes, _ int) lipgloss.Style {
+			return styles.SelectedNodeStyle
+		}
+	}
+	m.styles = styles
+	m.updateStyles()
+}
+
+// SetShowHelp shows or hides the help view.
+func (m *Model) SetShowHelp(v bool) {
+	m.showHelp = v
+}
+
+// SetSize sets the width and height of this component.
+func (m *Model) SetSize(width, height int) {
```

Comment:
Currently the `width` is applied only to the help model.
I saw other bubbles with width/height that also only apply them to select child models.
So I wasn't sure if:
1. Height should be applied anywhere
2. I need to wrap the `View()` return value in a style that applies them